File: src/TextFix/Core/FixService.cs
// SPDX-License-Identifier: GPL-3.0-only
using System.Net.NetworkInformation;
using TextFix.Ai;
using TextFix.Configuration;
using TextFix.Editing;
using TextFix.Interop;
using TextFix.Ui;

namespace TextFix.Core;

/// <summary>
/// Orchestrates one text fix: validate, read the text out of the focused app, run the request off
/// the message thread, then paste the result back. A second press while a request is in flight
/// cancels it rather than queuing another one.
///
/// Everything that touches the clipboard or synthesises keys runs on the message thread, so there
/// is exactly one writer and no lock needed around the target state.
///
/// The clipboard is the one piece of irreplaceable user data this code holds, so the rule around it
/// is absolute: a restore that has been armed is always settled before anything else is allowed to
/// touch the clipboard. <see cref="FlushPendingRestore"/> is that guarantee, and every path that
/// could otherwise drop a pending restore calls it - a second fix, a cancel, and shutdown.
/// </summary>
internal sealed class FixService
{
    internal enum State
    {
        Idle,
        Working,
    }

    internal const nuint RestoreTimerId = 21;

    /// <summary>Shown while the capture is running, before the request has gone out.</summary>
    private const string ReadingTooltip = "TextFix - reading text...";

    /// <summary>
    /// Shown while the request is in flight. It names the cancel gesture because a second press
    /// cancelling is otherwise undiscoverable - there is nowhere else the agent could say it.
    /// </summary>
    private const string WorkingTooltip = "TextFix - fixing text... (press the hotkey again to cancel)";

    private readonly nint _hwnd;
    private readonly TrayIcon _tray;
    private readonly Func<AppConfig> _config;

    private State _state = State.Idle;
    private long _token;
    private CancellationTokenSource? _cts;

    // Handed from the request thread to the message thread, read only after the completion
    // message for the matching token arrives.
    private TextCapture? _capture;
    private volatile string? _resultText;
    private volatile string? _errorText;
    private volatile string? _zdrFallbackModel;
    private ClipboardSnapshot? _pendingRestore;

    /// <summary>True while an indicator is on screen, so every exit path can take it down again.</summary>
    private bool _indicatorShown;

    internal FixService(nint hwnd, TrayIcon tray, Func<AppConfig> config)
    {
        _hwnd = hwnd;
        _tray = tray;
        _config = config;
    }

    internal void Start(int actionIndex)
    {
        // A second press is a cancel, not a second request: the first one still owns the caret.
        if (_state == State.Working)
        {
            Cancel();
            return;
        }

        // The previous fix may still have a clipboard restore sitting on the timer. Taking a
        // snapshot now would record our own output as "what the user was holding", and arming a
        // second restore would silently drop the first - so settle the old one before anything
        // here is allowed near the clipboard.
        FlushPendingRestore();

        AppConfig config = _config();
        if (actionIndex < 0 || actionIndex >= config.Actions.Count) return;
        FixActionConfig action = config.Actions[actionIndex];
        if (!action.Enabled)
        {
            Notify("Action disabled", "Enable it in TextFix settings.", isError: false);
            return;
        }

        AiProvider provider = config.ProviderValue;
        string apiKey = string.Empty;
        string model = string.Empty;
        if (!config.DryRun)
        {
            apiKey = SecretStore.GetApiKey(provider);
            if (apiKey.Length == 0)
            {
                Notify("No API key", "Add your " + provider.DisplayName() + " key in TextFix settings.", isError: true);
                return;
            }
            string? resolved = config.ResolveModel(action);
            if (resolved == null)
            {
                Notify("No model selected", "Pick a model in TextFix settings.", isError: true);
                return;
            }
            model = resolved;
            if (!IsNetworkAvailable())
            {
                Notify("Offline", "TextFix needs a connection to fix text.", isError: true);
                return;
            }

            // The request is committed from here on, so overlap the TLS handshake with the
            // clipboard round trip below instead of charging it to the user's wait.
            AiClient.Prewarm(provider);
        }

        // Feedback *before* the capture rather than after it. The whole-field path spends the
        // better part of a second in clipboard round trips, and a hotkey that changes nothing on
        // screen for that long reads as a dead key. The indicator starts in its quiet form -
        // present, not sweeping - and is promoted once the request is actually in flight, so the
        // two questions the user has at those two moments ("did my key register?" and "is it
        // stuck?") each get their own answer.
        nint foreground = Win32.GetForegroundWindow();
        ShowIndicator(config, TextTarget.GetCaretRect(foreground), foreground, sweeping: false);

        var options = new CaptureOptions(
            AllowSelectAll: action.WholeTextWhenNoSelection,
            SkipPasswordFields: config.SkipPasswordFields,
            TakeSnapshot: config.RestoreClipboard,
            SelectionProbeMs: config.SelectionProbeMs,
            CopyTimeoutMs: config.CopyTimeoutMs,
            // One character past the limit is all we ever need to read: the check below rejects
            // anything that long, and the cap is what stops a 200 MB clipboard being turned into
            // a string on the message thread just to discover it was never going to fit.
            MaxChars: config.MaxInputLength + 1);

        if (!TextTarget.TryCapture(_hwnd, options, out TextCapture? capture, out string captureError) || capture == null)
        {
            // TryCapture has already put the clipboard back; it is the only code that knows how
            // far the probe got.
            HideIndicator();
            Notify("Nothing fixed", captureError, isError: true);
            return;
        }

        if (capture.Text.Length > config.MaxInputLength)
        {
            HideIndicator();
            RestoreClipboardNow(capture.Saved);
            Notify("Too much text", "That is more than " + config.MaxInputLength + " characters.", isError: true);
            return;
        }

        _capture = capture;
        _resultText = null;
        _errorText = null;
        _zdrFallbackModel = null;
        _state = State.Working;
        long token = ++_token;

        // Now that the real caret rectangle is known, settle the indicator onto it and promote it
        // to the sweeping form. Repositioning does not restart the entrance animation.
        ShowIndicator(config, capture.CaretRect, capture.Foreground, sweeping: true);

        if (config.DryRun)
        {
            // Nothing leaves the machine: an obviously-transformed result proves the capture and
            // paste halves work in this app, which is the part that varies between hosts.
            _resultText = capture.Text.ToUpperInvariant();
            Log.Info("Dry run: " + capture.Text.Length + " characters, mode " + capture.Mode);
            Win32.PostMessage(_hwnd, WM.FIX_COMPLETE, (nuint)token, 0);
            return;
        }

        var request = new AiRequest(
            ApiKey: apiKey,
            Model: model,
            SystemPrompt: string.IsNullOrWhiteSpace(action.Prompt) ? Defaults.FixPrompt : action.Prompt.Trim(),
            Provider: provider,
            UseZeroDataRetention: config.ZeroDataRetention,
            DisableReasoning: !config.AllowReasoning,
            TotalBudgetMs: config.RequestBudgetMs);

        var cts = new CancellationTokenSource();
        _cts = cts;
        string input = capture.Text;

        _ = Task.Run(async () =>
        {
            try
            {
                AiResult result = await AiClient.FixTextAsync(input, request, cts.Token).ConfigureAwait(false);
                _resultText = result.Text;
                if (result.FellBackFromZdr) _zdrFallbackModel = request.Model;
            }
            catch (OperationCanceledException)
            {
                _errorText = null; // the user asked for it; nothing to report
            }
            catch (Exception e)
            {
                Log.Warn("Text fix failed: " + AiErrors.Scrub(e.GetType().Name + ": " + e.Message));
                _errorText = AiErrors.UserFacing(e, "Could not fix that text.");
            }
            finally
            {
                // Hand control back to the message thread; it owns the clipboard and the caret.
                Win32.PostMessage(_hwnd, WM.FIX_COMPLETE, (nuint)token, 0);
            }
        });
    }

    internal void Cancel()
    {
        if (_state != State.Working) return;
        _token++; // invalidates the in-flight completion message
        _cts?.Cancel();
        _cts = null;
        _state = State.Idle;
        HideIndicator();
        // Belt and braces: Start already settles any pending restore before it reaches Working,
        // so this should find nothing. It is here so the invariant survives a future edit that
        // reorders the two.
        FlushPendingRestore();
        RestoreClipboardNow(_capture?.Saved);
        _capture = null;
    }

    /// <summary>Runs on the message thread once the request thread has posted its result.</summary>
    internal void Complete(long token)
    {
        if (token != _token || _state != State.Working) return; // stale or cancelled

        _state = State.Idle;
        _cts = null;
        HideIndicator();

        TextCapture? capture = _capture;
        _capture = null;
        AppConfig config = _config();

        if (_zdrFallbackModel is string fallbackModel && AiClient.ShouldWarnAboutZdrFallback(fallbackModel))
        {
            Notify(
                "Zero data retention unavailable",
                "This request used standard routing for " + fallbackModel + ".",
                isError: false);
        }

        if (_errorText is string error)
        {
            RestoreClipboardNow(capture?.Saved);
            Notify("Text fix failed", error, isError: true);
            return;
        }
        if (_resultText == null || capture == null)
        {
            // Cancelled between the post and here.
            RestoreClipboardNow(capture?.Saved);
            return;
        }

        // Unwrap first: a prompt that fences its input in tags sometimes gets tags back.
        string proposed = AiText.Sanitize(AiText.Unwrap(_resultText), Defaults.MaxOutputLength);
        _resultText = null;
        if (proposed.Length == 0)
        {
            RestoreClipboardNow(capture.Saved);
            Notify("Text fix failed", "The model returned nothing usable.", isError: true);
            return;
        }
        if (string.Equals(proposed, capture.Text, StringComparison.Ordinal))
        {
            // Nothing to change. Pasting identical text would still cost the user an undo step
            // and a scroll jump, so leave the field alone.
            RestoreClipboardNow(capture.Saved);
            Log.Debug("Model returned the input unchanged; field left untouched");
            return;
        }

        ReplaceOutcome outcome = TextTarget.Replace(_hwnd, capture, proposed, config.PasteOnlyIfFocusUnchanged);
        switch (outcome)
        {
            case ReplaceOutcome.Pasted:
                if (capture.Saved != null)
                {
                    // The paste needs the clipboard to stay put for a moment: some hosts read it
                    // asynchronously after Ctrl+V returns. Restoring on a timer keeps the message
                    // thread free in the meantime.
                    _pendingRestore = capture.Saved;
                    Win32.SetTimer(_hwnd, RestoreTimerId, (uint)Math.Max(config.ClipboardRestoreDelayMs, 1), 0);
                }
                if (capture.Saved is { Complete: false })
                {
                    Notify(
                        "Clipboard partly restored",
                        "Some clipboard content could not be copied back.",
                        isError: false);
                }
                break;

            case ReplaceOutcome.FocusChanged:
                // Never paste into a window the user switched to. Hand them the text instead -
                // flagged transient, because a rewrite they never asked to keep has no business
                // being archived in Win+V history or synced to their other devices.
                ClipboardBridge.SetText(proposed, _hwnd, transient: true);
                Notify(
                    "Focus changed",
                    capture.Saved is { IsEmpty: false }
                        ? "The fixed text is on your clipboard, replacing what was there."
                        : "The fixed text is on your clipboard.",
                    isError: false);
                break;

            case ReplaceOutcome.Blocked:
                RestoreClipboardNow(capture.Saved);
                Notify(
                    "Windows blocked the paste",
                    "That window is running as administrator, so TextFix cannot type into it.",
                    isError: true);
                break;

            default:
                RestoreClipboardNow(capture.Saved);
                Notify("Text fix failed", "Could not paste the fixed text.", isError: true);
                break;
        }
    }

    internal void OnRestoreTimer() => FlushPendingRestore();

    /// <summary>
    /// Runs a clipboard restore that is still sitting on the timer, right now.
    ///
    /// Every caller exists because the restore would otherwise be lost. A second fix would
    /// overwrite the snapshot and reset the timer; a quit would drop the WM_TIMER entirely, since
    /// GetMessage retrieves WM_QUIT well ahead of WM_TIMER in the queue's priority order. Either
    /// way the user's clipboard would be gone and our own output left sitting in its place.
    /// </summary>
    internal void FlushPendingRestore()
    {
        ClipboardSnapshot? snapshot = _pendingRestore;
        if (snapshot == null) return;
        _pendingRestore = null;
        Win32.KillTimer(_hwnd, RestoreTimerId);
        ClipboardBridge.RestoreOrClear(snapshot, _hwnd);
    }

    private void RestoreClipboardNow(ClipboardSnapshot? snapshot)
    {
        if (snapshot != null) ClipboardBridge.RestoreOrClear(snapshot, _hwnd);
    }

    private void ShowIndicator(AppConfig config, RECT caret, nint foreground, bool sweeping)
    {
        if (config.Indicator == "caret") CaretPulse.Show(caret, foreground, sweeping);
        _indicatorShown = true;
        _tray.SetTooltip(sweeping ? WorkingTooltip : ReadingTooltip);
    }

    private void HideIndicator()
    {
        if (!_indicatorShown) return;
        _indicatorShown = false;
        CaretPulse.Hide();
        _tray.SetTooltip(AgentWindow.IdleTooltip);
    }

    private void Notify(string title, string message, bool isError)
    {
        if (isError) Log.Warn(title + ": " + message);
        else Log.Info(title + ": " + message);
        if (_config().NotifyOnError) _tray.ShowBalloon(title, message, isError);
    }

    private static bool IsNetworkAvailable()
    {
        try
        {
            return NetworkInterface.GetIsNetworkAvailable();
        }
        catch
        {
            // If the check itself fails, let the request decide rather than blocking the feature.
            return true;
        }
    }

    internal void Dispose()
    {
        _cts?.Cancel();
        _cts = null;
    }
}

File: src/TextFixMac/Sources/TextFix/Core/FixService.swift
// SPDX-License-Identifier: GPL-3.0-only
import AppKit

/// Orchestrates one text fix: validate, read the text out of the focused app, run the request off
/// the main thread, then write the result back. A second press while a request is in flight cancels
/// it rather than queuing another one.
///
/// Everything that touches the pasteboard, the Accessibility API or the keyboard runs on the main
/// thread, so there is exactly one writer and no lock around the target state. That confinement is
/// what `@unchecked Sendable` below asserts: the request task hops back through the main queue
/// before it touches anything here, and the compiler cannot see that for itself.
///
/// The pasteboard is the one piece of irreplaceable user data this code holds, so the rule around
/// it is absolute: a restore that has been scheduled is always settled before anything else is
/// allowed to touch the pasteboard. `flushPendingRestore` is that guarantee, and every path that
/// could otherwise drop one calls it - a second fix, a cancel, and the app quitting.
final class FixService {
    private enum State {
        case idle
        case working
    }

    private let config: () -> AppConfig
    private weak var statusItem: StatusItem?

    private var state = State.idle
    private var token = 0
    private var currentTask: Task<Void, Never>?
    private var capture: TextCapture?
    private var pendingRestore: PasteboardSnapshot?
    private var pendingRestoreWork: DispatchWorkItem?

    /// True while an indicator is on screen, so every exit path can take it down again.
    private var indicatorShown = false

    init(config: @escaping () -> AppConfig, statusItem: StatusItem) {
        self.config = config
        self.statusItem = statusItem
    }

    // ---- start ----

    func start(actionIndex: Int) {
        // A second press is a cancel, not a second request: the first one still owns the caret.
        if state == .working {
            cancel()
            return
        }

        // The previous fix may still have a pasteboard restore waiting on its timer. Taking a
        // snapshot now would record our own output as "what the user was holding", and scheduling
        // a second restore would silently drop the first - so settle the old one before anything
        // here is allowed near the pasteboard.
        flushPendingRestore()

        let config = self.config()
        guard actionIndex >= 0, actionIndex < config.actions.count else { return }
        let action = config.actions[actionIndex]
        guard action.enabled else {
            notify("Action disabled", "Enable it in TextFix settings.", isError: false)
            return
        }

        guard TextTarget.isTrusted else {
            // The one failure worth interrupting the user for, because it is the only one they can
            // fix and the only one that makes every hotkey press do nothing.
            Permissions.promptForAccessibility()
            return
        }

        let provider = config.providerValue
        var apiKey = ""
        var model = ""
        if !config.dryRun {
            apiKey = SecretStore.apiKey(for: provider)
            guard !apiKey.isEmpty else {
                notify("No API key", "Add your \(provider.displayName) key in TextFix settings.", isError: true)
                return
            }
            guard let resolved = config.resolveModel(action) else {
                notify("No model selected", "Pick a model in TextFix settings.", isError: true)
                return
            }
            model = resolved

            // The request is committed from here on, so overlap the TLS handshake with the capture
            // below instead of charging it to the user's wait.
            AiClient.prewarm(provider)
        }

        let options = CaptureOptions(
            allowSelectAll: action.wholeTextWhenNoSelection,
            skipPasswordFields: config.skipPasswordFields,
            takeSnapshot: config.restorePasteboard,
            useAccessibilityRead: config.useAccessibilityRead,
            selectionProbeMs: config.selectionProbeMs,
            copyTimeoutMs: config.copyTimeoutMs,
            // One code unit past the limit is all that is ever worth reading: the check below
            // rejects anything that long, and the cap is what stops a 200 MB pasteboard being
            // turned into a String on the main thread just to discover it was never going to fit.
            maxUTF16: config.maxInputLength + 1)

        let captured: TextCapture
        // Feedback goes up as soon as the target is known, rather than after the capture. The
        // pasteboard path spends the better part of a second in round trips, and a hotkey that
        // changes nothing on screen for that long reads as a dead key. It starts in its quiet form
        // - present, not sweeping - and is promoted once the request is actually in flight.
        switch TextTarget.capture(options: options, onTargetResolved: { [weak self] caretRect in
            self?.showIndicator(config: config, caretRect: caretRect, sweeping: false)
        }) {
        case .failed(let message):
            // capture() has already put the pasteboard back; it is the only code that knows how far
            // the probe got.
            hideIndicator()
            notify("Nothing fixed", message, isError: true)
            return
        case .captured(let result):
            captured = result
        }

        // Counted in UTF-16 code units, which is what the Windows agent counts, so one
        // maxInputLength in config.json means the same thing on both platforms.
        guard captured.text.utf16.count <= config.maxInputLength else {
            hideIndicator()
            restorePasteboardNow(captured.saved)
            notify("Too much text", "That is more than \(config.maxInputLength) characters.", isError: true)
            return
        }

        capture = captured
        state = .working
        token += 1
        let token = self.token

        // Now that the real caret rectangle is known, settle the indicator onto it and promote it
        // to the sweeping form. Repositioning does not restart the entrance animation.
        showIndicator(config: config, caretRect: captured.caretRect, sweeping: true)

        if config.dryRun {
            // Nothing leaves the machine: an obviously-transformed result proves the capture and
            // replace halves work in this app, which is the part that varies between hosts.
            Log.info("Dry run: \(captured.text.utf16.count) characters, mode \(captured.mode)")
            complete(token: token, text: captured.text.uppercased(), zdrFallbackModel: nil, error: nil)
            return
        }

        let request = AiRequest(
            apiKey: apiKey,
            model: model,
            systemPrompt: action.prompt.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
                ? Defaults.fixPrompt
                : action.prompt.trimmingCharacters(in: .whitespacesAndNewlines),
            provider: provider,
            useZeroDataRetention: config.zeroDataRetention,
            disableReasoning: !config.allowReasoning,
            totalBudgetMs: config.requestBudgetMs)

        let input = captured.text
        currentTask = Task { [weak self] in
            do {
                let result = try await AiClient.fixText(input, request)
                DispatchQueue.main.async {
                    self?.complete(
                        token: token,
                        text: result.text,
                        zdrFallbackModel: result.fellBackFromZdr ? request.model : nil,
                        error: nil)
                }
            } catch is CancellationError {
                // The user asked for it; nothing to report.
            } catch {
                Log.warn("Text fix failed: " + AiErrors.scrub("\(type(of: error)): \(error)"))
                let message = AiErrors.userFacing(error, fallback: "Could not fix that text.")
                DispatchQueue.main.async {
                    self?.complete(token: token, text: nil, zdrFallbackModel: nil, error: message)
                }
            }
        }
    }

    // ---- cancel ----

    func cancel() {
        guard state == .working else { return }
        token += 1 // invalidates the in-flight completion
        currentTask?.cancel()
        currentTask = nil
        state = .idle
        hideIndicator()
        // Belt and braces: start already settles any pending restore before it reaches .working,
        // so this should find nothing. It is here so the invariant survives a future edit that
        // reorders the two.
        flushPendingRestore()
        restorePasteboardNow(capture?.saved)
        capture = nil
    }

    // ---- finish ----

    private func complete(token: Int, text: String?, zdrFallbackModel: String?, error: String?) {
        guard token == self.token, state == .working else { return } // stale or cancelled

        state = .idle
        currentTask = nil
        hideIndicator()

        let capture = self.capture
        self.capture = nil
        let config = self.config()

        if let zdrFallbackModel, AiClient.shouldWarnAboutZdrFallback(zdrFallbackModel) {
            notify(
                "Zero data retention unavailable",
                "This request used standard routing for \(zdrFallbackModel).",
                isError: false)
        }

        if let error {
            restorePasteboardNow(capture?.saved)
            notify("Text fix failed", error, isError: true)
            return
        }
        guard let text, let capture else {
            // Cancelled between the request finishing and here.
            restorePasteboardNow(capture?.saved)
            return
        }

        // Unwrap first: a prompt that fences its input in tags sometimes gets tags back.
        let proposed = AiText.sanitize(AiText.unwrap(text), maxLength: Defaults.maxOutputLength)
        guard !proposed.isEmpty else {
            restorePasteboardNow(capture.saved)
            notify("Text fix failed", "The model returned nothing usable.", isError: true)
            return
        }
        guard proposed != capture.text else {
            // Nothing to change. Writing identical text would still cost the user an undo step and
            // a scroll jump, so leave the field alone.
            restorePasteboardNow(capture.saved)
            Log.debug("Model returned the input unchanged; field left untouched")
            return
        }

        let outcome = TextTarget.replace(
            capture: capture,
            with: proposed,
            requireSameFocus: config.replaceOnlyIfFocusUnchanged,
            mode: config.replaceModeValue)

        switch outcome {
        case .replaced:
            // Whether the pasteboard was borrowed is a fact, not something to infer from the
            // configured mode: an Accessibility write that was refused falls back to a paste, and
            // a paste that happened needs the delay whatever asked for it.
            let pasteboardMoved = capture.saved.map { $0.changeCount != PasteboardBridge.changeCount } ?? false
            if let saved = capture.saved, pasteboardMoved {
                // The paste needs the pasteboard to stay put for a moment: some hosts read it
                // asynchronously after Cmd+V returns.
                schedulePasteboardRestore(saved, afterMs: config.pasteboardRestoreDelayMs)
            } else {
                restorePasteboardNow(capture.saved)
            }
            if capture.saved?.complete == false {
                notify(
                    "Clipboard partly restored",
                    "Some clipboard content could not be copied back.",
                    isError: false)
            }

        case .focusChanged:
            // Never write into a window the user switched to. Hand them the text instead - flagged
            // transient, because a rewrite they never asked to keep has no business being archived
            // by a clipboard manager or synced to their other devices.
            PasteboardBridge.setText(proposed, transient: true)
            notify(
                "Focus changed",
                capture.saved.map { !$0.isEmpty } == true
                    ? "The fixed text is on your clipboard, replacing what was there."
                    : "The fixed text is on your clipboard.",
                isError: false)

        case .blocked:
            restorePasteboardNow(capture.saved)
            Permissions.promptForAccessibility()

        case .failed:
            restorePasteboardNow(capture.saved)
            notify("Text fix failed", "Could not put the fixed text back.", isError: true)
        }
    }

    // ---- pasteboard ----

    private func schedulePasteboardRestore(_ snapshot: PasteboardSnapshot, afterMs: Int) {
        flushPendingRestore()
        pendingRestore = snapshot
        let work = DispatchWorkItem { [weak self] in
            guard let self else { return }
            self.pendingRestoreWork = nil
            self.flushPendingRestore()
        }
        pendingRestoreWork = work
        DispatchQueue.main.asyncAfter(deadline: .now() + Double(max(afterMs, 1)) / 1000, execute: work)
    }

    /// Runs a pasteboard restore that is still waiting on its timer, right now.
    ///
    /// Every caller exists because the restore would otherwise be lost. A second fix would
    /// overwrite the snapshot; quitting would cancel the queued block outright. Either way the
    /// user's clipboard would be gone and our own output left sitting in its place.
    func flushPendingRestore() {
        pendingRestoreWork?.cancel()
        pendingRestoreWork = nil
        guard let snapshot = pendingRestore else { return }
        pendingRestore = nil
        PasteboardBridge.restoreOrClear(snapshot)
    }

    private func restorePasteboardNow(_ snapshot: PasteboardSnapshot?) {
        guard let snapshot else { return }
        PasteboardBridge.restoreOrClear(snapshot)
    }

    // ---- shared ----

    private func showIndicator(config: AppConfig, caretRect: CGRect?, sweeping: Bool) {
        switch config.indicatorValue {
        case .caret:
            CaretPulse.shared.show(caretRect: caretRect, sweeping: sweeping)
        case .menubar:
            statusItem?.setWorking(true)
        case .none:
            break
        }
        indicatorShown = true
        // The tooltip is the only place the agent can say that a second press cancels, which is
        // otherwise completely undiscoverable.
        statusItem?.setTooltip(sweeping
            ? "TextFix - fixing text… (press the hotkey again to cancel)"
            : "TextFix - reading text…")
    }

    private func hideIndicator() {
        guard indicatorShown else { return }
        indicatorShown = false
        CaretPulse.shared.hide()
        statusItem?.setWorking(false)
        statusItem?.setTooltip(StatusItem.idleTooltip)
    }

    private func notify(_ title: String, _ message: String, isError: Bool) {
        if isError { Log.warn("\(title): \(message)") } else { Log.info("\(title): \(message)") }
        guard config().notifyOnError else { return }
        Toast.shared.show(title: title, message: message, isError: isError)
    }

    func shutdown() {
        currentTask?.cancel()
        currentTask = nil
        // Settled rather than discarded: a quit inside the restore delay would otherwise leave the
        // user holding our output instead of whatever they had copied.
        flushPendingRestore()
    }
}

/// Main-thread-confined by construction - see the type comment. The request task captures it only
/// to hop back through the main queue, which is the one thing the compiler cannot verify for
/// itself, so the guarantee is asserted here rather than re-established with a lock.
extension FixService: @unchecked Sendable {}
